3p-ecosystem/src/main/java/com/google/homesampleapp/screens/thread/ThreadNetworkUiState.kt

```diff
@@ -0,0 +1,268 @@
+import com.google.android.gms.common.api.CommonStatusCodes.getStatusCodeString
+import com.google.android.gms.threadnetwork.ThreadBorderAgent
+import com.google.android.gms.threadnetwork.ThreadNetwork
+import com.google.android.gms.threadnetwork.ThreadNetworkCredentials
+import com.google.android.gms.threadnetwork.ThreadNetworkStatusCodes
+import com.google.common.io.BaseEncoding
+import com.google.mlkit.vision.barcode.common.Barcode
+import com.google.mlkit.vision.codescanner.GmsBarcodeScannerOptions
+import com.google.mlkit.vision.codescanner.GmsBarcodeScanning
+import timber.log.Timber
+
+/**
+ * Implements UI logic associated with ThreadFragment.
+ * That UI logic is mainly associated with API calls that require Activity.
+ * Actions performed by the user in the Fragment trigger calls to processAction(),
+ * and the impact of theses actions are communicated back to the UI via StateFlows.
+ */
+class ThreadNetworkUiState(
+  private val activity: ComponentActivity,
+  private val viewModel: ThreadViewModel
+) {
+
+  private lateinit var sd: ServiceDiscovery
+
+  // -----------------------------------------------------------------------------------------------
+  // Process Action triggered from the UI
+
+  fun processAction(actionRequest: ActionRequest) {
+    Timber.d("processAction [${actionRequest.type}] [${actionRequest.task}] [${actionRequest.serviceInfo}]")
+
+    // Dispatch processing according to action type.
+    when (actionRequest.type) {
+      ActionType.None -> {
+        // No currently active action. No Action Dialog
+        viewModel.setActionDialogInfo(ActionType.None, ActionState.None)
+      }
+
+      ActionType.doGpsPreferredCredentialsExist -> {
+        doGPSPreferredCredsExist(actionRequest)
+      }
+
+      ActionType.getGpsPreferredCredentials -> {
+        getGPSPreferredCreds(actionRequest)
+      }
+
+      ActionType.setGpsPreferredCredentials -> {
+        setGPSThreadCredentials(actionRequest)
+      }
+
+      ActionType.clearGpsPreferredCredentials -> {
+        viewModel.clearGPSPreferredCreds(actionRequest)
+      }
+
+      ActionType.getOtbrActiveThreadCredentials -> {
+        viewModel.getOTBRActiveThreadCredentials(actionRequest, getThreadBorderRoutersList())
+      }
+
+      ActionType.setOtbrPendingThreadCredentials -> {
+        viewModel.setOTBRPendingThreadCredentials(actionRequest, getThreadBorderRoutersList())
+      }
+
+      ActionType.readQrCodeCredentials -> {
+        readQRCodeWorkingSet(actionRequest)
+      }
+
+      ActionType.showQrCodeCredentials -> {
+        viewModel.showWorkingSetQRCode(actionRequest)
+      }
+    }
+  }
+
+  // -----------------------------------------------------------------------------------------------
+  // Actions for GPS Preferred Credentials
+
+  /**
+   * Checks whether credentials exist in storage or not. Consent from user is not necessary.
+   * Requires Activity.
+   */
+  private fun doGPSPreferredCredsExist(actionRequest: ActionRequest) {
```

Comment:
This method and `getGPSPreferredCreds`, `setGPSThreadCredentials` and other methods that use the `Task` API and its requiste callbacks may be easier to manage by converting them to suspending methods and using an [`await`](https://github.com/Kotlin/kotlinx.coroutines/blob/master/integration/kotlinx-coroutines-play-services/src/Tasks.kt#L106) call on the `Task`

Reply:
Yes. great suggestion. Added a TODO. Will address once the migration has completed.